### PR TITLE
chore: coverage reporting + backend CRS-logging and type hardening

### PR DESCRIPTION
**Pull request by giswqs**
Implements three review follow-ups: coverage measurement, backend type hints/docstrings, and surfacing silent CRS swallows.

## 1. Coverage measurement (non-blocking)
- Add `test:frontend:coverage` (Node's built-in `--experimental-test-coverage`, zero new dependency) and `test:backend:coverage` (`pytest-cov`).
- CI now runs the `:coverage` variants, so every build prints a coverage summary. It still gates on test failures but has **no coverage threshold yet** (informational first; ratchet a floor once a baseline exists).
- Add `pytest-cov` to the backend `dev` extra; ignore coverage artifacts in `.gitignore`.
- Chose the Node built-in over `c8` because `c8` crashes on Node 26 (a yargs ESM bug); the built-in works on both Node 26 and CI's Node 22. Caveat documented in `CLAUDE.md`: the frontend report only counts files a test imports, so a fully-untested module is absent rather than shown as 0%.

Baselines from this run: frontend **79.6% lines**; backend **59% total** (`vector_ops.py` 92%, previously-invisible `whitebox.py`/`sedona_ops.py` now at 30%/20%).

## 2. Backend `vector_ops.py` type hints + docstrings
- Add `Optional[dict]` / `dict[str, Any]` parameter type hints to all 19 tool handlers (previously bare `geojson, overlay, parameters`) and type `_to_feature_collection(gdf: Any)`.
- Add Google-style docstrings to the 13 handlers that lacked them plus the two `_smooth_*` helpers.

## 3. Silent CRS swallows
- **`raster.py`** (3 sites): these live inside embedded subprocess scripts whose stdout/stderr become visible job messages, so a swallowed CRS-parse failure was invisible. Replaced `except: pass` / silent fallback with a `print(...warning..., file=sys.stderr)` that surfaces as a job message while keeping the WGS84 fallback.
- **`ml.py`** (2 sites): parent-process code, so added `logger.debug(...)` for the health-probe failure and the SIGTERM->SIGKILL escalation, with justifying `# noqa: BLE001`.

## Validation
- Frontend: 847 pass / 0 fail with coverage. Backend: 192 passed / 6 skipped with coverage.
- `ruff check` clean on changed files; embedded raster scripts confirmed to still `compile()`.
- Full `pre-commit run --all-files` passes (including the npm build).

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved health-check and process-termination probe handling with clearer debug logging when failures occur.
  * Added clearer CRS parsing warnings during raster vector operations, while preserving the existing EPSG:4326 fallback.
* **Refactor/Style**
  * Improved vector-operation type annotations and expanded related documentation (behavior unchanged).
* **Tests**
  * Added coverage-focused frontend and backend test scripts and updated CI to run them.
* **Documentation**
  * Updated test instructions to explain coverage report contents/omissions and threshold behavior.
* **Chores**
  * Ignored Python coverage artifacts and added coverage tooling to the dev dependencies.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -21,7 +21,9 @@
     "lint": "eslint apps packages workers tests",
     "test": "npm run test:frontend",
     "test:frontend": "node --import tsx --test tests/*.test.ts",
+    "test:frontend:coverage": "node --import tsx --test --experimental-test-coverage --test-coverage-exclude='tests/**' --test-coverage-exclude='e2e/**' --test-coverage-exclude='**/*.config.*' tests/*.test.ts",
```
[the remark this answers is not in the document]
Comment (giswqs) [reply package.json]: Done in 01b5573 — switched the `--test-coverage-exclude` globs to escaped double quotes so they group correctly under Windows cmd.exe.

[the remark this answers is not in the document]
Comment (giswqs) [reply package.json]: Done in 01b5573 — updated the `ci` script to call `test:frontend:coverage` and `test:backend:coverage`, so `npm run ci` now has parity with GitHub CI and exercises the coverage flags locally.

File: backend/geolibre_server/geolibre_server/app/raster.py [hunk 2]
```diff
@@ -819,8 +828,12 @@ def model(hh):
         if digits:
             try:
                 zone_crs = CRS.from_epsg(int(digits.group(1)))
-            except Exception:
-                pass
+            except Exception as exc:
```
[the remark this answers is not in the document]
Comment (giswqs) [reply backend/geolibre_server/geolibre_server/app/raster.py:844]: Done in 01b5573 — added an `else` branch on the zonal script's outer `except` so a CRS name with no trailing digits also emits a warning instead of silently staying WGS84. Used an else (rather than warning in the outer except directly) to avoid a spurious warning when the digit-parse fallback succeeds.

File: backend/geolibre_server/geolibre_server/app/ml.py [hunk 3]
```diff
@@ -241,7 +242,8 @@ def _terminate_process(proc) -> None:
     proc.terminate()
     try:
         proc.wait(timeout=5)
-    except Exception:
+    except Exception as exc:  # noqa: BLE001 - escalate to kill on any wait failure
+        logger.debug("Process ignored SIGTERM (%s); sending SIGKILL", exc)
```
[the remark this answers is not in the document]
Comment (giswqs) [reply backend/geolibre_server/geolibre_server/app/ml.py]: Done in e1d5cfd — reworded to "SIGTERM wait failed (%s); escalating to SIGKILL" so it's accurate for any wait failure, not just TimeoutExpired.

File: backend/geolibre_server/geolibre_server/app/raster.py [hunk 4]
```diff
@@ -593,7 +597,12 @@ class RasterToolRequest(BaseModel):
             pass
 try:
     crs = CRS.from_epsg(epsg)
-except Exception:
+except Exception as exc:
+    print(
```
[the remark this answers is not in the document]
Comment (giswqs) [reply backend/geolibre_server/geolibre_server/app/raster.py:614]: Done in e1d5cfd — removed the unreachable `try/except ValueError` around `int(digits.group(1))` (digits matches `(\d+)$`, so int() can't raise) and left a one-line comment in its place.

File: backend/geolibre_server/geolibre_server/vector_ops.py [hunk 5]
```diff
@@ -81,15 +81,24 @@ def _load_gdf(geojson: Optional[dict], label: str) -> Any:
...
     """Serialize a GeoDataFrame back to a GeoJSON FeatureCollection dict."""
     # GeoPandas only emits valid GeoJSON in WGS84; reproject if needed.
     if gdf.crs is not None and gdf.crs.to_epsg() != 4326:
         gdf = gdf.to_crs(WGS84)
     return json.loads(gdf.to_json())
 
 
-def _buffer(geojson, overlay, parameters) -> tuple[dict, list[str]]:
+def _buffer(
+    geojson: Optional[dict], overlay: Optional[dict], parameters: dict[str, Any]
```
[the remark this answers is not in the document]
Comment (giswqs) [reply backend/geolibre_server/geolibre_server/vector_ops.py:97]: Done in e1d5cfd — documented the shared `(geojson, overlay, parameters)` signature once in a comment above the first handler, noting single-layer tools ignore `overlay`, rather than repeating it across every single-layer docstring.

File: backend/geolibre_server/geolibre_server/app/ml.py [hunk 6]
```diff
@@ -133,7 +133,8 @@ def _is_healthy(base_url: str, timeout: float = 3.0) -> bool:
         # The `ml` extra (httpx) is missing: report "not reachable" rather than
         # raising, so callers like /ml/status degrade gracefully.
         return False
-    except Exception:
+    except Exception as exc:  # noqa: BLE001 - a probe failure means "not healthy"
+        logger.debug("Health check for %s failed: %s", base_url, exc)
```
[the remark this answers is not in the document]
Comment (giswqs) [reply backend/geolibre_server/geolibre_server/app/ml.py]: Done in a056254 — wrapped `base_url` in `_redact_url()` so credentials in GEOLIBRE_ML_SAMGEO_URL can't leak into the debug log, matching the existing redaction at 169/309/317. Good catch.

File: backend/geolibre_server/geolibre_server/vector_ops.py [hunk 7]
```diff
@@ -184,7 +214,10 @@ def _overlay_op(geojson, overlay, parameters, how: str) -> tuple[dict, list[str]
     )
 
 
-def _clip(geojson, overlay, parameters) -> tuple[dict, list[str]]:
+def _clip(
+    geojson: Optional[dict], overlay: Optional[dict], parameters: dict[str, Any]
+) -> tuple[dict, list[str]]:
+    """Clip the input layer to the overlay layer's extent."""
```
[the remark this answers is not in the document]
Comment (giswqs) [reply backend/geolibre_server/geolibre_server/vector_ops.py]: Done in a056254 — changed the `_clip` docstring from "extent" to "geometry" since gpd.clip cuts to the overlay's actual polygon boundary, not its bounding box.

File: backend/geolibre_server/geolibre_server/app/ml.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (giswqs) [reply backend/geolibre_server/geolibre_server/app/ml.py]: Already addressed in e1d5cfd — the message no longer says "Process ignored SIGTERM"; it now reads "SIGTERM wait failed (%s); escalating to SIGKILL", which is neutral for any wait() failure (and includes exc). This re-flag describes the prior wording. Resolving as already fixed.

File: package.json [hunk 8]
```diff
@@ -21,11 +21,13 @@
...
     "test": "npm run test:frontend",
     "test:frontend": "node --import tsx --test tests/*.test.ts",
+    "test:frontend:coverage": "node --import tsx --test --experimental-test-coverage --test-coverage-exclude=\"tests/**\" --test-coverage-exclude=\"e2e/**\" --test-coverage-exclude=\"**/*.config.*\" tests/*.test.ts",
     "test:backend": "python -m pytest backend/geolibre_server/tests",
+    "test:backend:coverage": "python -m pytest backend/geolibre_server/tests --cov=geolibre_server --cov-report=term-missing",
     "test:worker": "npm run typecheck -w geolibre-viewer-worker && npm run typecheck -w geolibre-collab-worker",
     "test:e2e": "playwright test",
     "check:rust": "cargo check --manifest-path apps/geolibre-desktop/src-tauri/Cargo.toml",
-    "ci": "npm run lint && npm run build && npm run test:frontend && npm run test:worker && npm run test:backend && npm run check:rust",
+    "ci": "npm run lint && npm run build && npm run test:frontend:coverage && npm run test:worker && npm run test:backend:coverage && npm run check:rust",
```
[the remark this answers is not in the document]
Comment (giswqs) [reply package.json:30]: Done in 9fd4ba3 — added a note to CLAUDE.md's Tests section: the backend coverage run and `npm run ci` need pytest-cov from the backend `dev` extra (`pip install -e "backend/geolibre_server[dev]"`), analogous to the playwright-install callout.

[the remark this answers is not in the document]
Comment (giswqs) [reply package.json:30]: Done in 9fd4ba3 — kept `ci` on the coverage variants (parity, as requested last round) and documented the pytest-cov requirement in CLAUDE.md so the `unrecognized arguments: --cov` failure mode is covered by an explicit install note.

File: backend/geolibre_server/geolibre_server/app/raster.py [hunk 9]
```diff
@@ -819,8 +826,18 @@ def model(hh):
...
                 zone_crs = CRS.from_epsg(int(digits.group(1)))
-            except Exception:
-                pass
+            except Exception as exc:
+                print(
+                    f"Warning: could not parse zones CRS '{name}' "
+                    f"({exc}); assuming WGS84 (EPSG:4326)",
+                    file=sys.stderr,
+                )
+        else:
```
[the remark this answers is not in the document]
Comment (giswqs) [reply backend/geolibre_server/geolibre_server/app/raster.py]: Done in 9fd4ba3 — changed the zones `else` to `elif name:` so an absent/blank crs.properties.name no longer prints the misleading "could not parse zones CRS ''" warning.

File: backend/geolibre_server/geolibre_server/app/raster.py [hunk 10]
```diff
@@ -333,8 +333,12 @@ class RasterToolRequest(BaseModel):
...
         try:
             mask_crs = CRS.from_epsg(int(digits.group(1)))
-        except Exception:
-            pass
+        except Exception as exc:
+            print(
+                f"Warning: could not parse mask CRS '{name}' "
+                f"({exc}); assuming WGS84 (EPSG:4326)",
+                file=sys.stderr,
+            )
```
[the remark this answers is not in the document]
Comment (giswqs) [reply backend/geolibre_server/geolibre_server/app/raster.py:341]: Done in 9fd4ba3 — added matching `elif name:` warnings to the mask section (and the interpolate section at ~line 595) so all three embedded scripts surface an uninterpretable CRS name consistently and stay silent for a missing field.

File: backend/geolibre_server/geolibre_server/app/raster.py [hunk 11]
```diff
@@ -819,8 +826,18 @@ def model(hh):
...
+                    f"Warning: could not parse zones CRS '{name}' "
+                    f"({exc}); assuming WGS84 (EPSG:4326)",
+                    file=sys.stderr,
+                )
+        else:
+            print(
+                f"Warning: could not parse zones CRS '{name}'; "
+                "assuming WGS84 (EPSG:4326)",
+                file=sys.stderr,
+            )
```
[the remark this answers is not in the document]
Comment (giswqs) [reply backend/geolibre_server/geolibre_server/app/raster.py]: Leaving this open for a maintainer call rather than churning another round. Two points:

1. Partially stale: as of 9fd4ba3 the `else:` is now `elif name:` (it no longer fires for an absent/blank name).

2. Intentional trade-off: I kept this no-match branch detail-free so all three embedded scripts (mask, interpolate, zones) emit the **same** shape of warning. Mask and interpolate only do digit extraction (no `from_user_input` call), so they have no equivalent exception to report. Capturing `from_user_input`'s exception only in the zones branch would make zones' warning richer than the other two and re-introduce the cross-section inconsistency the previous review round asked me to remove. The message already names the offending CRS string, which is the actionable part.

Happy to add the `(exc)` detail to the zones branch if you'd prefer detail over cross-section symmetry.
